app/sso.js

```diff
@@ -0,0 +1,45 @@
+const crypto = require('crypto');
+const fs = require('fs');
+
+function generateToken() {
+	// Normally we would use 16 => 128 bits of entropy which is sufficiennt
+	// But since we're going to base64 it and there would be padding (==),
+	// It's a wasted space, why not use padding for some additional entropy? :)
+	// The replacing is to make it URL-safe
+	return crypto.randomBytes(18).toString("base64").replace('+', '-').replace('/', '_')
+}
+
+function updateToken(tokenFile) {
+	var newToken = generateToken();
+	var tmpFileName = tokenFile + ".tmp";
+	fs.writeFileSync(tmpFileName, newToken);
+	fs.renameSync(tmpFileName, tokenFile);
+	return newToken;
+}
+
+module.exports = (tokenFile, loginRedirect) => {
+	var token = updateToken(tokenFile)
+	var cookies = new Set()
+
+	return (req, res, next) => {
+		if (req.cookies && cookies.has(req.cookies.btcexp_auth)) {
+			req.authenticated = true;
+			return next();
+		}
+
+		if (req.query.token === token) {
+			req.authenticated = true;
+			token = updateToken(tokenFile);
```

Comment:
Does this actually use the token written to the file? It seems like the next line ignores that and creates a new token instead.

Maybe instead of updating the file every time the token is used, it could use HMAC authentication with a counter or a recent timestamp?

Reply:
It generates a new token and saves it into a file on start and simultaneously keeps it in memory. Then on each login compares the token to the one in memory and if it matches, it regenerates the secret again. This is the same flow RTL and ThunderHub do. I was thinking about a few other ways too but came to conclusion that this approach is simple (both for consumer and provider) and since other services do it, the code can be reused.

I'm aware that there's a race condition if two users login at the same time, I don't think it's worth spending our time on (private nodes don't have hundreds of users per second logging in). Of course, I specifically made my SSO provider to read the token when a user is actually attempting to open the app, not beforehand.

Reply:
I agree with the goal of simplicity @Kixunil and am fine with what we have currently implemented (thanks for the context about other projects methodology in this area since I don't use this feature). That said, I'm open to other ideas @shesek if the tradeoffs warrant a change.

Reply:
If my judgement about race condition being improbable is wrong, I will be the one to make a fix. :)